relay-sampling/src/evaluation.rs

```diff
@@ -95,139 +114,193 @@ pub fn merge_configs_and_match(
-    {
-        let mut matched_rule_ids = vec![];
-        // Even though this seed is changed based on whether we match event or trace rules, we will
-        // still incur in inconsistent trace sampling because of multi-matching of rules across event
-        // and trace rules.
-        //
-        // An example of inconsistent trace sampling could be:
-        // /hello -> /world -> /transaction belong to trace_id = abc
-        // * /hello has uniform rule with 0.2 sample rate which will match all the transactions of the trace
-        // * each project has a single transaction rule with different factors (2, 3, 4)
-        //
-        // 1. /hello is matched with a transaction rule with a factor of 2 and uses as seed abc -> 0.2 * 2 = 0.4 sample rate
-        // 2. /world is matched with a transaction rule with a factor of 3 and uses as seed abc -> 0.2 * 3 = 0.6 sample rate
-        // 3. /transaction is matched with a transaction rule with a factor of 4 and uses as seed abc -> 0.2 * 4 = 0.8 sample rate
-        //
-        // We can see that we have 3 different samples rates but given the same seed, the random number generated will be the same.
-        let mut seed = event.and_then(|e| e.id.value()).map(|id| id.0);
-        let mut accumulated_factors = 1.0;
-
-        for rule in rules {
-            let matches = match rule.ty {
-                RuleType::Trace => match dsc {
-                    Some(dsc) => rule.condition.matches(dsc),
-                    _ => false,
-                },
-                RuleType::Transaction => event.map_or(false, |event| match event.ty.0 {
-                    Some(EventType::Transaction) => rule.condition.matches(event),
-                    _ => false,
-                }),
+    /// Returns true if the event should be kept.
+    pub fn should_keep(&self) -> bool {
+        match self {
+            SamplingResult::Match { is_kept, .. } => *is_kept,
+            // If no rules matched on an event, we want to keep it.
+            SamplingResult::NoMatch => true,
+        }
+    }
+
+    /// Returns true if the event should be dropped.
+    pub fn should_drop(&self) -> bool {
+        !self.should_keep()
+    }
+}
+
+/// Matches an event and/or dynamic sampling context against the rules of the sampling configuration.
+///
+/// The multi-matching algorithm used iterates by collecting and multiplying factor rules until
+/// it finds a sample rate rule. Once a sample rate rule is found, the final sample rate is
+/// computed by multiplying it with the previously accumulated factors.
+///
+/// The default accumulated factors equal to 1 because it is the identity of the multiplication
+/// operation, thus in case no factor rules are matched, the final result will just be the
+/// sample rate of the matching rule.
+///
+/// In case no sample rate rule is matched, we are going to return a None, signaling that no
+/// match has been found.
+pub(crate) fn get_sampling_match<'a>(
```

Comment:
pub(crate) only due to test, i can look into moving stuff around to get rid of it

Reply:
update: got rid of it by moving the tests